src/storage/src/hummock/event_handler/uploader.rs

```diff
@@ -445,44 +428,291 @@ impl SealedData {
+    }
+
+    fn local_seal_epoch(
+        &mut self,
+        instance_id: LocalInstanceId,
+        next_epoch: HummockEpoch,
+        opts: SealCurrentEpochOptions,
+    ) {
+        let instance_data = self
+            .instance_data
+            .get_mut(&instance_id)
+            .expect("should exist");
+        let epoch = instance_data.local_seal_epoch(next_epoch);
+        self.epoch_data.entry(next_epoch).or_default();
+        if let Some((direction, table_watermarks)) = opts.table_watermarks {
+            self.epoch_data
+                .get_mut(&epoch)
+                .expect("should be created in a previous init_epoch or local_seal_epoch")
+                .add_table_watermarks(instance_data.table_id, table_watermarks, direction);
+        }
+    }
+
+    fn may_destroy_instance(&mut self, instance_id: LocalInstanceId) {
+        if let Some(instance) = self.instance_data.remove(&instance_id) {
+            let instance_ids = self
+                .table_instance_id
+                .get_mut(&instance.table_id)
+                .expect("should exist");
+            assert!(instance_ids.remove(&instance_id));
+            if instance_ids.is_empty() {
+                assert!(self.table_instance_id.remove(&instance.table_id).is_some());
+            }
+        }
+    }
+
+    fn may_spill(&mut self, context: &UploaderContext) {
+        if context.buffer_tracker.need_more_flush() {
+            // iterate from older epoch to newer epoch
+            for (epoch, epoch_data) in &mut self.epoch_data {
+                let mut payload = HashMap::new();
+                for (instance_id, instance_data) in &mut self.instance_data {
+                    let instance_payload = instance_data.spill(*epoch);
+                    if !instance_payload.is_empty() {
+                        payload.insert(*instance_id, instance_payload);
+                    }
+                }
+                epoch_data.flush(context, payload);
+                if !context.buffer_tracker.need_more_flush() {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn sync(
+        &mut self,
+        epoch: HummockEpoch,
+    ) -> (HashMap<LocalInstanceId, Vec<ImmutableMemtable>>, SealedData) {
+        let mut sync_epoch_data = self.epoch_data.split_off(&(epoch + 1));
+        swap(&mut sync_epoch_data, &mut self.epoch_data);
+
+        // by now, epoch data inclusively before epoch is in sync_epoch_data
+
+        let mut sealed_data = SealedData::default();
+        for (epoch, epoch_data) in sync_epoch_data {
+            sealed_data.seal_new_epoch(epoch, epoch_data);
+        }
+
+        let mut flush_payload = HashMap::new();
+        for (instance_id, instance_data) in &mut self.instance_data {
+            let payload = instance_data.sync(epoch);
+            if !payload.is_empty() {
+                flush_payload.insert(*instance_id, payload);
+            }
+        }
+        (flush_payload, sealed_data)
+    }
```

Comment:
`epoch_data` and `instance_data` are independently added to `sealed_data` while the same epoch can be both in `epoch_data` and `instance_data`. Also, `epoch_data` is added via `sealed_data.seal_new_epoch` in the `sync` call while `instance_data` is added via `sealed_data.flush` immediately after the `sync` call. 

Will it be clearer if we move the data from `instance_data` to `epoch_data` before updating `sealed_data`, similar to what we did in `may_spill`? For example, in `UnsyncData::sync`, we can do:
1. iterate through all instance data, call `instance_data.sync` and collect `payload`.
2. call `epoch_data.flush` to add payload to `epoch_data`.
3. split_off `epoch_data` and init `sealed_data`.
4. return a single `sealed_data`.